File: ClassEval_T/cpp/solution/BalancedBrackets.cpp

#include <algorithm>

#include <stack>
#include <string>

class BalancedBrackets {
public:
    explicit BalancedBrackets(const std::string& expr);
    bool check_balanced_brackets();

    void clearExpr();

    std::stack<char> stack;
    const std::string leftBrackets = "({[";
    const std::string rightBrackets = ")}]";
    std::string expr;
};

BalancedBrackets::BalancedBrackets(const std::string& expr) : expr(expr) {}

void BalancedBrackets::clearExpr() {
    expr.erase(std::remove_if(expr.begin(), expr.end(), [this](char c) {
        return leftBrackets.find(c) == std::string::npos && rightBrackets.find(c) == std::string::npos;
        }), expr.end());
}

bool BalancedBrackets::check_balanced_brackets() {
    clearExpr();
    for (char Brkt : expr) {
        if (leftBrackets.find(Brkt) != std::string::npos) {
            stack.push(Brkt);
        }
        else {
            if (stack.empty()) return false;
            char Current_Brkt = stack.top();
            stack.pop();
            if (Current_Brkt == '(' && Brkt != ')') return false;
            if (Current_Brkt == '{' && Brkt != '}') return false;
            if (Current_Brkt == '[' && Brkt != ']') return false;
        }
    }
    return stack.empty();
}

File: BatchTestTool/java/java automated tester/Test/compile_passed/solutions/BalancedBrackets.java
import java.util.Stack;
import java.util.ArrayList;
import java.util.List;

public class BalancedBrackets {
    private Stack<Character> stack;
    private List<Character> leftBrackets;
    private List<Character> rightBrackets;
    private String expr;

    public BalancedBrackets(String expr) {
        this.stack = new Stack<>();
        this.leftBrackets = new ArrayList<>(List.of('(', '{', '['));
        this.rightBrackets = new ArrayList<>(List.of(')', '}', ']'));
        this.expr = expr;
    }

    public void clearExpr() {
        StringBuilder sb = new StringBuilder();
        for (char c : expr.toCharArray()) {
            if (leftBrackets.contains(c) || rightBrackets.contains(c)) {
                sb.append(c);
            }
        }
        this.expr = sb.toString();
    }

    public boolean checkBalancedBrackets() {
        clearExpr();
        for (char Brkt : expr.toCharArray()) {
            if (leftBrackets.contains(Brkt)) {
                stack.push(Brkt);
            } else {
                char Current_Brkt = stack.pop();
                if (Current_Brkt == '(') {
                    if (Brkt != ')') {
                        return false;
                    }
                } else if (Current_Brkt == '{') {
                    if (Brkt != '}') {
                        return false;
                    }
                } else if (Current_Brkt == '[') {
                    if (Brkt != ']') {
                        return false;
                    }
                }
            }
        }
        return stack.isEmpty();
    }
}